diff --git a/kernel-work/linux-6.18.33/include/net/netfilter/nf_conntrack.h b/kernel-work/linux-6.18.33/include/net/netfilter/nf_conntrack.h
index 87b3edbc8..2cf144233 100644
--- a/kernel-work/linux-6.18.33/include/net/netfilter/nf_conntrack.h
+++ b/kernel-work/linux-6.18.33/include/net/netfilter/nf_conntrack.h
@@ -112,6 +112,7 @@ struct nf_conn {
         u_int32_t mark;
         u_int32_t ipset_mark;
 #endif
+        int multi_gw_ifindex;
 
 #ifdef CONFIG_NF_CONNTRACK_SECMARK
         u_int32_t secmark;
diff --git a/kernel-work/linux-6.18.33/net/netfilter/nf_nat_core.c b/kernel-work/linux-6.18.33/net/netfilter/nf_nat_core.c
index 1aee29672..42e24f0ed 100644
--- a/kernel-work/linux-6.18.33/net/netfilter/nf_nat_core.c
+++ b/kernel-work/linux-6.18.33/net/netfilter/nf_nat_core.c
@@ -15,6 +15,7 @@
 #include <net/xfrm.h>
 #include <linux/siphash.h>
 #include <linux/rtnetlink.h>
+#include <linux/inetdevice.h>
 #include <net/ip.h>
 #include <net/route.h>
 
@@ -1327,34 +1328,53 @@ unsigned int ip_nat_route_input(void *priv,
         enum ip_conntrack_dir dir;
         unsigned long statusbit;
         __be32 saddr;
+        struct net_device *device;
+        struct net *net;
 
         if (!(conn = nf_ct_get(skb, &ctinfo)))
                 return NF_ACCEPT;
 
-        if (!(conn->status & IPS_NAT_DONE_MASK))
-                return NF_ACCEPT;
+        if (!(conn->status & IPS_NAT_DONE_MASK)) {
+                if (!(conn->multi_gw_ifindex))
+                        return NF_ACCEPT;
+
+                net = dev_net(skb->dev);
+                device = dev_get_by_index(net, conn->multi_gw_ifindex);
+
+                if (device != NULL) {
+                        __be32 out_addr = inet_select_addr(device, 0, RT_SCOPE_LINK);
+                        iph = ip_hdr(skb);
+                        dev_put(device);
+                        if (ip_route_input_lookup(skb, iph->daddr, iph->saddr,
+                            ip4h_dscp(iph), skb->dev, out_addr))
+                                return NF_DROP;
+                }
+        }
+
         dir = CTINFO2DIR(ctinfo);
         statusbit = IPS_SRC_NAT;
         if (dir == IP_CT_DIR_REPLY)
                 statusbit ^= IPS_NAT_MASK;
-        if (!(conn->status & statusbit))
-                return NF_ACCEPT;
 
-        if (skb_dst(skb))
+        if (!(conn->status & statusbit)) {
                 return NF_ACCEPT;
+        } else {
+                if (skb_dst(skb))
+                        return NF_ACCEPT;
 
-        if (skb->len < sizeof(struct iphdr))
-                return NF_ACCEPT;
+                if (skb->len < sizeof(struct iphdr))
+                        return NF_ACCEPT;
 
-        /* use daddr in other direction as masquerade address (lsrc) */
-        iph = ip_hdr(skb);
-        saddr = conn->tuplehash[!dir].tuple.dst.u3.ip;
-        if (saddr == iph->saddr)
-                return NF_ACCEPT;
+                /* use daddr in other direction as masquerade address (lsrc) */
+                iph = ip_hdr(skb);
+                saddr = conn->tuplehash[!dir].tuple.dst.u3.ip;
+                if (saddr == iph->saddr)
+                        return NF_ACCEPT;
 
-        if (ip_route_input_lookup(skb, iph->daddr, iph->saddr,
-            ip4h_dscp(iph), skb->dev, saddr))
-                return NF_DROP;
+                if (ip_route_input_lookup(skb, iph->daddr, iph->saddr,
+                    ip4h_dscp(iph), skb->dev, saddr))
+                        return NF_DROP;
+        }
 
         return NF_ACCEPT;
 }
